core/drivers/tpm2/tpm2_ptp_fifo.c

```diff
@@ -0,0 +1,477 @@
+
+	/* locality will be relinquishd in tpm2_rx() */
+	ret = tpm2_fifo_request_locality(chip, 0);
+	if (ret)
+		return ret;
+
+	while (retry_cnt < TPM2_DD_RETRY_CNT) {
+		ret = TPM2_OK;
+		buf_len = len;
+		sent = 0;
+		wr_size = 0;
+		burstcnt = 0;
+		status = 0;
+
+		/*
+		 * If unable to get status, something fundamental is
+		 * wrong. No retries needed.
+		 */
+		ret = tpm2_fifo_get_status(chip, &status);
+		if (ret)
+			break;
+
+		if (!(status & TPM2_STS_COMMAND_READY)) {
+			/*
+			 * If unable to set status, something fundamental is
+			 * wrong. No retries needed.
+			 */
+			ret = tpm2_fifo_set_status(chip,
+						   TPM2_STS_COMMAND_READY);
+			if (ret) {
+				EMSG("Previous cmd cancel failed");
+				break;
+			}
+			ret = tpm2_fifo_wait_for_status(chip,
+							TPM2_STS_COMMAND_READY,
+							chip->timeout_b,
+							&status);
+			if (ret) {
+				retry_cnt++;
+				continue;
+			}
+		}
+
+		while (buf_len > 0) {
+			ret = tpm2_fifo_get_burstcount(chip, &burstcnt);
+			if (ret)
+				break;
+
+			wr_size = MIN(buf_len, burstcnt);
+			ret = ops->tx8(chip, TPM2_DATA_FIFO(chip->locality),
+				       wr_size, buf + sent);
+			if (ret)
+				break;
+
+			ret = tpm2_fifo_wait_for_status(chip, TPM2_STS_VALID,
+							chip->timeout_a,
+							&status);
+			if (ret)
+				break;
+
+			sent += wr_size;
+			buf_len -= wr_size;
+			/* TPM2 should expect more data */
+			if (buf_len && !(status & TPM2_STS_DATA_EXPECT)) {
+				ret = TPM2_ERR_IO;
+				break;
+			}
+		}
+
+		/* If any error has occurred in transmit loop above retry */
+		if (ret) {
+			retry_cnt++;
+			continue;
+		}
+
+		/* last check everything is ok and TPM2 expects no more data */
+		ret = tpm2_fifo_wait_for_status(chip, TPM2_STS_VALID,
+						chip->timeout_a, &status);
+		if (ret) {
+			retry_cnt++;
+			continue;
+		}
+
+		/* All data has been read, TPM2 should not expect more data */
+		if (status & TPM2_STS_DATA_EXPECT) {
+			retry_cnt++;
+			continue;
+		}
+
+		ret = tpm2_fifo_set_status(chip, TPM2_STS_GO);
+		if (ret)
+			break;
+
+		return TPM2_OK;
+	}
+
+	/* This will result in current command getting cancelled */
+	tpm2_fifo_set_status(chip, TPM2_STS_COMMAND_READY);
+	/* We are here because some error occurred, so relinquish locality */
+	tpm2_fifo_relinquish_locality(chip);
+
+	return ret;
+}
+
+/* Based on Fig 4 - Receive response using FIFO protocol in DD Specification */
+enum tpm2_result tpm2_fifo_recv(struct tpm2_chip *chip, uint8_t *buf,
+				uint32_t *len, uint32_t cmd_duration)
+{
+	uint32_t burstcnt = 0;
+	uint32_t bytes2read = TPM2_HDR_LEN;
+	uint32_t param_size_flag = 0;
+	uint32_t sz = 0;
+	uint32_t rxsize = 0;
+	uint8_t retry_cnt = 0;
+	uint8_t status = 0;
+	uint8_t flags = 0;
+	enum tpm2_result ret = TPM2_OK;
+	const struct tpm2_ptp_phy_ops *ops = NULL;
+
+	if (!chip)
+		return TPM2_ERR_GENERIC;
+
+	ops = chip->phy_ops;
+
+	if (*len < TPM2_HDR_LEN) {
+		EMSG("Unable to read TPM2 header\n");
+		*len = TPM2_HDR_LEN;
+		return TPM2_ERR_INVALID_ARG;
```

Comment:
An inline comment should describe that output `len` is updated on `TPM2_ERR_INVALID_ARG` .
I even think a specific error code is needed here, to distinguish from other places where the function may return `TPM2_ERR_INVALID_ARG`.

Reply:
Ok..Will add a new error code TPM2_ERR_SHORT_BUFFER.